pkg/controller/database/dbsubnetgroup/controller.go

```diff
@@ -18,175 +18,221 @@ package dbsubnetgroup
-	if len(response.DBSubnetGroups) != 1 {
-		return managed.ExternalObservation{}, errors.Errorf(errMultipleItems, cr.Spec.DBSubnetGroupName)
+	if len(res.DBSubnetGroups) != 1 {
+		return managed.ExternalObservation{}, errors.Errorf(errZeroOrMoreResouce, meta.GetExternalName(cr))
 	}
 
-	observed := response.DBSubnetGroups[0]
-
-	cr.SetConditions(runtimev1alpha1.Available())
-
-	cr.UpdateExternalStatus(observed)
+	observed := res.DBSubnetGroups[0]
+	current := cr.Spec.ForProvider.DeepCopy()
+	dbsg.LateInitialize(&cr.Spec.ForProvider, &observed)
+	if !reflect.DeepEqual(current, &cr.Spec.ForProvider) {
+		if err := e.kube.Update(ctx, cr); err != nil {
+			return managed.ExternalObservation{}, errors.Wrap(err, errKubeUpdateFailed)
+		}
+	}
+	cr.Status.AtProvider = dbsg.GenerateObservation(observed)
+
+	switch cr.Status.AtProvider.State {
+	case v1beta1.DBSubnetGroupStateAvailable:
+		cr.Status.SetConditions(runtimev1alpha1.Available())
+		resource.SetBindable(cr)
+	case v1beta1.DBSubnetGroupStateCreating:
+		cr.Status.SetConditions(runtimev1alpha1.Creating())
+	case v1beta1.DBSubnetGroupStateDeleting:
+		cr.Status.SetConditions(runtimev1alpha1.Deleting())
+	default:
+		cr.Status.SetConditions(runtimev1alpha1.Unavailable())
+	}
 
 	return managed.ExternalObservation{
-		ResourceExists:    true,
-		ConnectionDetails: managed.ConnectionDetails{},
+		ResourceUpToDate: dbsg.IsDBSubnetGroupUpToDate(cr.Spec.ForProvider, observed),
+		ResourceExists:   true,
 	}, nil
 }
 
 func (e *external) Create(ctx context.Context, mgd resource.Managed) (managed.ExternalCreation, error) {
-	cr, ok := mgd.(*v1alpha3.DBSubnetGroup)
+	cr, ok := mgd.(*v1beta1.DBSubnetGroup)
 	if !ok {
 		return managed.ExternalCreation{}, errors.New(errUnexpectedObject)
 	}
 
-	cr.Status.SetConditions(runtimev1alpha1.Creating())
+	cr.SetConditions(runtimev1alpha1.Creating())
 
-	input := &awsrds.CreateDBSubnetGroupInput{
-		DBSubnetGroupDescription: aws.String(cr.Spec.DBSubnetGroupDescription),
-		DBSubnetGroupName:        aws.String(cr.Spec.DBSubnetGroupName),
-		SubnetIds:                cr.Spec.SubnetIDs,
-		Tags:                     []awsrds.Tag{},
+	if cr.Status.AtProvider.State == v1beta1.DBSubnetGroupStateCreating {
+		return managed.ExternalCreation{}, nil
 	}
 
-	for _, t := range cr.Spec.Tags {
-		input.Tags = append(input.Tags, awsrds.Tag{
-			Key:   aws.String(t.Key),
-			Value: aws.String(t.Value),
-		})
+	input := &awsrds.CreateDBSubnetGroupInput{
+		DBSubnetGroupDescription: aws.String(cr.Spec.ForProvider.Description),
+		DBSubnetGroupName:        aws.String(meta.GetExternalName(cr)),
+		SubnetIds:                cr.Spec.ForProvider.SubnetIDs,
 	}
 
-	req := e.client.CreateDBSubnetGroupRequest(input)
+	if len(cr.Spec.ForProvider.Tags) != 0 {
+		input.Tags = make([]awsrds.Tag, len(cr.Spec.ForProvider.Tags))
+		for i, val := range cr.Spec.ForProvider.Tags {
+			input.Tags[i] = awsrds.Tag{Key: &val.Key, Value: &val.Value}
```

Comment:
```suggestion
			input.Tags[i] = awsrds.Tag{Key: aws.String(val.Key), Value: aws.String(val.Value)}
```

`&val.Key` refers to the address of `val.Key` string. The value of `val.Key` changes at each iteration of the for loop but its address doesn't actually change. What you end up with is that all elements of `input.Tags` pointing to the same address, whose value was assigned in the last iteration of that for loop. So, you get an array full of the same values, which results in `DuplicateKey` error. I've encountered a similar issue [here in RDS](https://github.com/crossplane/provider-aws/pull/178).